From: Jiri Slaby <jslaby@suse.cz>
Subject: x86-non-upstream-eager-fpu 32bit fix
Patch-mainline: never, SUSE specific
References: bnc#1087086 CVE-2018-3665

This is a fix for:
patches.suse/x86-non-upstream-eager-fpu.patch

Due to missing commit 72a671ced66 (x86, fpu: Unify signal handling code
paths for x86 and x86_64 kernels), I omitted this 32bit path.

Signed-off-by: Jiri Slaby <jslaby@suse.cz>
---
 arch/x86/kernel/i387.c |    7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

--- a/arch/x86/kernel/i387.c
+++ b/arch/x86/kernel/i387.c
@@ -585,7 +585,7 @@ int save_i387_xstate_ia32(void __user *b
 	 * This will cause a "finit" to be triggered by the next
 	 * attempted FPU operation by the 'current' process.
 	 */
-	clear_used_math();
+	drop_init_fpu(tsk);
 
 	if (!HAVE_HWFP) {
 		return fpregs_soft_get(current, NULL,
@@ -682,10 +682,7 @@ int restore_i387_xstate_ia32(void __user
 		clear_fpu(tsk);
 
 	if (!buf) {
-		if (used_math()) {
-			clear_fpu(tsk);
-			clear_used_math();
-		}
+		drop_init_fpu(tsk);
 
 		return 0;
 	} else
